SelectMuxerUpgrade is located in the private builder module

### Summary

We used a [SelectUpgrade](https://github.com/fluencelabs/nox/blob/e0de88cc4ca21546a26bad703314dbee6235f5ff/crates/libp2p/src/transport.rs#L82) in our code to build a transport. Now we can't use a SelectUpgrade in the multiplex function anymore it requires InboundConnectionUpgrade, but it is not implemented for the SelectUpgrade. Instead of this, I tried to change it to SelectMuxerUpgrade, but it is private.

### Expected behavior

Successful compilation

### Actual behavior

Fails with
```
    Checking fluence-libp2p v0.2.0 (/Users/gurinderu/projects/fluencelabs/rust-peer/crates/libp2p)
error[E0277]: the trait bound `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>: InboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not satisfied
   --> crates/libp2p/src/transport.rs:90:20
    |
90  |         .multiplex(multiplex)
    |          --------- ^^^^^^^^^ the trait `InboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not implemented for `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>`
    |          |
    |          required by a bound introduced by this call
    |
    = help: the following other types implement trait `InboundConnectionUpgrade<T>`:
              <libp2p::libp2p_yamux::Config as InboundConnectionUpgrade<C>>
              <MplexConfig as InboundConnectionUpgrade<C>>
              <libp2p::libp2p_noise::Config as InboundConnectionUpgrade<T>>
              <libp2p::builder::select_muxer::SelectMuxerUpgrade<A, B> as InboundConnectionUpgrade<C>>
              <libp2p::builder::select_security::SelectSecurityUpgrade<A, B> as InboundConnectionUpgrade<C>>
note: required by a bound in `Authenticated::<T>::multiplex`
   --> /Users/gurinderu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libp2p-core-0.41.0/src/transport/upgrade.rs:239:12
    |
231 |     pub fn multiplex<C, M, U, E>(
    |            --------- required by a bound in this associated function
...
239 |         U: InboundConnectionUpgrade<Negotiated<C>, Output = M, Error = E>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Authenticated::<T>::multiplex`

error[E0277]: the trait bound `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>: OutboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not satisfied
   --> crates/libp2p/src/transport.rs:90:20
    |
90  |         .multiplex(multiplex)
    |          --------- ^^^^^^^^^ the trait `OutboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not implemented for `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>`
    |          |
    |          required by a bound introduced by this call
    |
    = help: the following other types implement trait `OutboundConnectionUpgrade<T>`:
              <libp2p::libp2p_yamux::Config as OutboundConnectionUpgrade<C>>
              <MplexConfig as OutboundConnectionUpgrade<C>>
              <libp2p::libp2p_noise::Config as OutboundConnectionUpgrade<T>>
              <libp2p::builder::select_muxer::SelectMuxerUpgrade<A, B> as OutboundConnectionUpgrade<C>>
              <libp2p::builder::select_security::SelectSecurityUpgrade<A, B> as OutboundConnectionUpgrade<C>>
note: required by a bound in `Authenticated::<T>::multiplex`
   --> /Users/gurinderu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libp2p-core-0.41.0/src/transport/upgrade.rs:240:12
    |
231 |     pub fn multiplex<C, M, U, E>(
    |            --------- required by a bound in this associated function
...
240 |         U: OutboundConnectionUpgrade<Negotiated<C>, Output = M, Error = E> + Clone,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Authenticated::<T>::multiplex`

error[E0277]: the trait bound `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>: InboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not satisfied
  --> crates/libp2p/src/transport.rs:92:10
   |
92 |         .boxed()
   |          ^^^^^ the trait `InboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not implemented for `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>`
   |
   = help: the following other types implement trait `InboundConnectionUpgrade<T>`:
             <libp2p::libp2p_yamux::Config as InboundConnectionUpgrade<C>>
             <MplexConfig as InboundConnectionUpgrade<C>>
             <libp2p::libp2p_noise::Config as InboundConnectionUpgrade<T>>
             <libp2p::builder::select_muxer::SelectMuxerUpgrade<A, B> as InboundConnectionUpgrade<C>>
             <libp2p::builder::select_security::SelectSecurityUpgrade<A, B> as InboundConnectionUpgrade<C>>
   = note: required for `Multiplex<Output<Negotiated<C>>, SelectUpgrade<Config, MplexConfig>>` to implement `futures::Future`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-9584121374718173012.txt'
   = note: required for `Multiplex<Output<Negotiated<C>>, SelectUpgrade<Config, MplexConfig>>` to implement `TryFuture`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-9584121374718173012.txt'
   = note: required for `AndThen<AndThen<T, impl (FnOnce(C, ConnectedPoint) -> Authenticate<C, Config>) + Clone>, ...>` to implement `libp2p::Transport`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-15198203099857116664.txt'
   = note: 1 redundant requirement hidden
   = note: required for `TransportTimeout<AndThen<AndThen<T, impl (FnOnce(C, ConnectedPoint) -> Authenticate<C, Config>) + Clone>, ...>>` to implement `libp2p::Transport`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-3060906632340275684.txt'

error[E0277]: the trait bound `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>: OutboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not satisfied
  --> crates/libp2p/src/transport.rs:92:10
   |
92 |         .boxed()
   |          ^^^^^ the trait `OutboundConnectionUpgrade<multistream_select::negotiated::Negotiated<libp2p::libp2p_noise::Output<multistream_select::negotiated::Negotiated<C>>>>` is not implemented for `SelectUpgrade<libp2p::libp2p_yamux::Config, MplexConfig>`
   |
   = help: the following other types implement trait `OutboundConnectionUpgrade<T>`:
             <libp2p::libp2p_yamux::Config as OutboundConnectionUpgrade<C>>
             <MplexConfig as OutboundConnectionUpgrade<C>>
             <libp2p::libp2p_noise::Config as OutboundConnectionUpgrade<T>>
             <libp2p::builder::select_muxer::SelectMuxerUpgrade<A, B> as OutboundConnectionUpgrade<C>>
             <libp2p::builder::select_security::SelectSecurityUpgrade<A, B> as OutboundConnectionUpgrade<C>>
   = note: required for `Multiplex<Output<Negotiated<C>>, SelectUpgrade<Config, MplexConfig>>` to implement `futures::Future`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-9584121374718173012.txt'
   = note: required for `Multiplex<Output<Negotiated<C>>, SelectUpgrade<Config, MplexConfig>>` to implement `TryFuture`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-9584121374718173012.txt'
   = note: required for `AndThen<AndThen<T, impl (FnOnce(C, ConnectedPoint) -> Authenticate<C, Config>) + Clone>, ...>` to implement `libp2p::Transport`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-15198203099857116664.txt'
   = note: 1 redundant requirement hidden
   = note: required for `TransportTimeout<AndThen<AndThen<T, impl (FnOnce(C, ConnectedPoint) -> Authenticate<C, Config>) + Clone>, ...>>` to implement `libp2p::Transport`
   = note: the full type name has been written to '/Users/gurinderu/projects/fluencelabs/rust-peer/target/debug/deps/fluence_libp2p-6cf6ff149f5c305d.long-type-3060906632340275684.txt'

For more information about this error, try `rustc --explain E0277`.
error: could not compile `fluence-libp2p` (lib) due to 4 previous errors
```

### Relevant log output

_No response_

### Possible Solution

SelectMuxerUpgrade is public or SelectUpgrade implements InboundConnectionUpgrade.

### Version

0.53.0

### Would you like to work on fixing this bug ?

Maybe

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -75,7 +75,7 @@ libp2p = { version = "0.53.0", path = "libp2p" }
 libp2p-allow-block-list = { version = "0.3.0", path = "misc/allow-block-list" }
 libp2p-autonat = { version = "0.12.0", path = "protocols/autonat" }
 libp2p-connection-limits = { version = "0.3.0", path = "misc/connection-limits" }
-libp2p-core = { version = "0.41.0", path = "core" }
+libp2p-core = { version = "0.41.1", path = "core" }
 libp2p-dcutr = { version = "0.11.0", path = "protocols/dcutr" }
 libp2p-dns = { version = "0.41.0", path = "transports/dns" }
 libp2p-floodsub = { version = "0.44.0", path = "protocols/floodsub" }

--- a/core/CHANGELOG.md
+++ b/core/CHANGELOG.md
@@ -1,6 +1,6 @@
 ## 0.41.1
 - Add  `{In,Out}boundConnectionUpgrade` for `SelectUpgrade`.
-  See [PR 4812](https://github.com/libp2p/rust-libp2p/pull/4695).
+  See [PR 4812].
 
 [PR 4812]: https://github.com/libp2p/rust-libp2p/pull/4812
 

--- a/core/CHANGELOG.md
+++ b/core/CHANGELOG.md
@@ -1,4 +1,4 @@
-## 0.41.1
+## 0.41.1 - unreleased
 - Add  `{In,Out}boundConnectionUpgrade` for `SelectUpgrade`.
   See [PR 4812].
 

--- a/core/CHANGELOG.md
+++ b/core/CHANGELOG.md
@@ -1,3 +1,9 @@
+## 0.41.1
+- Add  `{In,Out}boundConnectionUpgrade` for `SelectUpgrade`.
+  See [PR 4812](https://github.com/libp2p/rust-libp2p/pull/4695).
+
+[PR 4812]: https://github.com/libp2p/rust-libp2p/pull/4812
+
 ## 0.41.0
 
 - Remove blanket-impl of `{In,Out}boundUpgrade` for `{In,Out}boundConnectionUpgrade`.

--- a/core/CHANGELOG.md
+++ b/core/CHANGELOG.md
@@ -1,8 +1,7 @@
 ## 0.41.1 - unreleased
-- Add  `{In,Out}boundConnectionUpgrade` for `SelectUpgrade`.
-  See [PR 4812].
 
-[PR 4812]: https://github.com/libp2p/rust-libp2p/pull/4812
+- Implement `{In,Out}boundConnectionUpgrade` for `SelectUpgrade`.
+  See [PR 4812](https://github.com/libp2p/rust-libp2p/pull/4812).
 
 ## 0.41.0
 

--- a/core/Cargo.toml
+++ b/core/Cargo.toml
@@ -3,7 +3,7 @@ name = "libp2p-core"
 edition = "2021"
 rust-version = { workspace = true }
 description = "Core traits and structs of libp2p"
-version = "0.41.0"
+version = "0.41.1"
 authors = ["Parity Technologies <admin@parity.io>"]
 license = "MIT"
 repository = "https://github.com/libp2p/rust-libp2p"

--- a/core/src/upgrade/select.rs
+++ b/core/src/upgrade/select.rs
@@ -19,7 +19,10 @@
 // DEALINGS IN THE SOFTWARE.
 
 use crate::either::EitherFuture;
-use crate::upgrade::{InboundUpgrade, OutboundUpgrade, UpgradeInfo};
+use crate::upgrade::{
+    InboundConnectionUpgrade, InboundUpgrade, OutboundConnectionUpgrade, OutboundUpgrade,
+    UpgradeInfo,
+};
 use either::Either;
 use futures::future;
 use std::iter::{Chain, Map};
@@ -84,10 +87,44 @@ where
     }
 }
 
+impl<C, A, B, TA, TB, EA, EB> InboundConnectionUpgrade<C> for SelectUpgrade<A, B>
+where
+    A: InboundConnectionUpgrade<C, Output = TA, Error = EA>,
+    B: InboundConnectionUpgrade<C, Output = TB, Error = EB>,
+{
+    type Output = future::Either<TA, TB>;
+    type Error = Either<EA, EB>;
+    type Future = EitherFuture<A::Future, B::Future>;
+
+    fn upgrade_inbound(self, sock: C, info: Self::Info) -> Self::Future {
+        match info {
+            Either::Left(info) => EitherFuture::First(self.0.upgrade_inbound(sock, info)),
+            Either::Right(info) => EitherFuture::Second(self.1.upgrade_inbound(sock, info)),
+        }
+    }
+}
+
 impl<C, A, B, TA, TB, EA, EB> OutboundUpgrade<C> for SelectUpgrade<A, B>
 where
     A: OutboundUpgrade<C, Output = TA, Error = EA>,
     B: OutboundUpgrade<C, Output = TB, Error = EB>,
+{
+    type Output = future::Either<TA, TB>;
+    type Error = Either<EA, EB>;
+    type Future = EitherFuture<A::Future, B::Future>;
+
+    fn upgrade_outbound(self, sock: C, info: Self::Info) -> Self::Future {
+        match info {
+            Either::Left(info) => EitherFuture::First(self.0.upgrade_outbound(sock, info)),
+            Either::Right(info) => EitherFuture::Second(self.1.upgrade_outbound(sock, info)),
+        }
+    }
+}
+
+impl<C, A, B, TA, TB, EA, EB> OutboundConnectionUpgrade<C> for SelectUpgrade<A, B>
+where
+    A: OutboundConnectionUpgrade<C, Output = TA, Error = EA>,
+    B: OutboundConnectionUpgrade<C, Output = TB, Error = EB>,
 {
     type Output = future::Either<TA, TB>;
     type Error = Either<EA, EB>;

